File: mailblastr-java/src/main/java/com/mailblastr/resources/Batch.java
package com.mailblastr.resources;

import com.mailblastr.MailblastrResponse;
import com.mailblastr.http.ApiClient;
import com.mailblastr.requests.BatchEmailRequest;
import com.mailblastr.requests.SendEmailRequest;

import java.util.List;

/** Batch send — {@code mailblastr.batch().send(...)}. */
public final class Batch extends Resource {
    public Batch(ApiClient api) { super(api); }

    /**
     * Send up to 100 emails in one request. {@code POST /emails/batch}
     *
     * @deprecated use {@link #sendEmails(List)} — batch items reject
     *     {@code attachments} and {@code scheduled_at} (send those individually
     *     via {@code emails().send}), which {@link BatchEmailRequest} enforces
     *     at compile time.
     */
    @Deprecated
    public MailblastrResponse send(List<SendEmailRequest> requests) {
        return api.request("POST", "/emails/batch", requests);
    }

    /** @deprecated use {@link #sendEmails(List, String)} — see {@link #send(List)}. */
    @Deprecated
    public MailblastrResponse send(List<SendEmailRequest> requests, String idempotencyKey) {
        return api.request("POST", "/emails/batch", requests, idempotencyKey);
    }

    /**
     * Send up to 100 emails in one request. Batch items reject
     * {@code attachments} and {@code scheduled_at} — send those individually
     * via {@code emails().send}. {@code POST /emails/batch}
     *
     * <p><strong>Two success shapes, chosen by batch SIZE alone.</strong> Up to
     * 40 emails are sent while the request is open ({@code 200}); that body has
     * no {@code queued} key at all. A batch of 41–100 is ACCEPTED AND QUEUED
     * ({@code 202}): the body then carries {@code queued: true} and
     * {@code queued_count}, the ids are real ({@code emails().get(id)} works),
     * but every row starts at {@code scheduled} and NOTHING has been
     * transmitted — the worker sends on its next tick, so poll
     * {@code emails().get(id)} for the real outcome. A batch carrying any
     * {@code @mailblastr.dev} simulator recipient (in {@code to}, {@code cc} or
     * {@code bcc}) stays inline at any size.
     *
     * <pre>{@code
     * MailblastrResponse res = mailblastr.batch().sendEmails(requests);
     * if (res.statusCode() == 202) {                 // accepted, not sent
     *     Integer n = res.getInt("queued_count");    // null on the inline path
     * }
     * }</pre>
     *
     * <p>Branch on {@code res.statusCode() == 202} (or
     * {@code Boolean.TRUE.equals(res.getBoolean("queued"))}), never on a bare
     * {@code res.getBoolean("queued")}: it is a boxed {@code Boolean} and is
     * {@code null} for every inline batch, so unboxing it there throws.
     */
    public MailblastrResponse sendEmails(List<BatchEmailRequest> requests) {
        return api.request("POST", "/emails/batch", requests);
    }

    /**
     * Like {@link #sendEmails(List)}, with an {@code Idempotency-Key} so a
     * retry replays the first response instead of re-sending.
     *
     * <p>The key must be <strong>1–255 characters</strong> after trimming
     * (anything else is a {@code 400 invalid_idempotency_key}) and is bound to
     * the request body.
     *
     * <p>When a batch fails partway the error body always carries {@code sent}
     * (the ids already sent) and {@code sent_count}, with or without a key; read
     * them from {@link com.mailblastr.MailblastrException#getBody()}. The key
     * changes the STATUS, not the data: a keyless partial is downgraded to
     * {@code 422} precisely so no client auto-retries and re-sends the prefix.
     *
     * <p>A queued batch ({@code 202} — see {@link #sendEmails(List)} for when
     * size sends it down that path) records its answer under the key inside the
     * same transaction as the rows, so a retry replays the {@code 202} rather
     * than enqueuing a second copy.
     */
    public MailblastrResponse sendEmails(List<BatchEmailRequest> requests, String idempotencyKey) {
        return api.request("POST", "/emails/batch", requests, idempotencyKey);
    }
}

File: mailblastr-php/src/Resources/Batch.php
<?php

declare(strict_types=1);

namespace Mailblastr\Resources;

/**
 * Batch send — `$mailblastr->batch->send([...])`.
 */
class Batch extends Resource
{
    /**
     * Send up to 100 emails in one request. POST /emails/batch
     *
     * Batch items reject 'attachments' and 'scheduled_at' — send those
     * individually via `$mailblastr->emails->send()`.
     *
     * A batch is served one of two ways, chosen by SIZE alone, and BOTH are
     * success statuses — branch on the response, never on "no exception was
     * thrown":
     *  - 1-40 emails are sent INLINE (HTTP 200). The returned array has NO
     *    'queued' key at all — absent, never false. Every id in ['data'] has
     *    already been handed to the mail service.
     *  - 41-100 emails are QUEUED for the worker (HTTP 202). The array then
     *    also carries 'queued' => true and 'queued_count' (=== the number of
     *    ['data'] rows). The ids are real, but the emails are still
     *    'scheduled' and NOTHING has been transmitted yet; the worker sends
     *    them on its next tick. Test it with `$res['queued'] ?? false`.
     * A batch carrying an @mailblastr.dev simulator recipient (in to, cc or
     * bcc) stays inline at any size.
     *
     * @param array $payloads A list of send-email payloads.
     * @param array $options  'idempotencyKey' to safely retry.
     * @return array Always 'data'; plus 'queued' and 'queued_count' when queued.
     */
    public function send(array $payloads, array $options = []): array
    {
        return $this->client->request('POST', '/emails/batch', $payloads, $options);
    }
}
